File: 프로그래머스/레벨2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <vector>
#include <deque>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    deque<int> bridge(bridge_length, 0);
    int totalWeight = 0;
    int time = 0;
    
    for (int truck : truck_weights) {
        while (true) {
            time++;
            totalWeight -= bridge.front();
            bridge.pop_front();

            if (totalWeight + truck <= weight) {
                bridge.push_back(truck);
                totalWeight += truck;
                break;
            } else {
                bridge.push_back(0);
            }
        }
    }

    return time + bridge_length;
}

File: 프로그래머스/레벨2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    bridge = deque([0] * bridge_length)
    total_weight = 0
    time = 0
    
    for truck in truck_weights:
        while True:
            time += 1
            total_weight -= bridge.popleft()
            
            if total_weight + truck <= weight:
                bridge.append(truck)
                total_weight += truck
                break
            else:
                bridge.append(0)
    
    return time + bridge_length